verifiers/internal/gcb/provenance.go

```diff
@@ -0,0 +1,308 @@
+package gcb
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa01 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.1"
+	dsselib "github.com/secure-systems-lab/go-securesystemslib/dsse"
+
+	serrors "github.com/slsa-framework/slsa-verifier/errors"
+	"github.com/slsa-framework/slsa-verifier/options"
+	"github.com/slsa-framework/slsa-verifier/verifiers/internal/gcb/keys"
+)
+
+var GCBBuilderIDs = []string{"https://cloudbuild.googleapis.com/GoogleHostedWorker@v0.2"}
+
+type v01IntotoStatement struct {
+	intoto.StatementHeader
+	Predicate slsa01.ProvenancePredicate `json:"predicate"`
+}
+
+type gloudProvenance struct {
+	ImageSummary struct {
+		Digest              string `json:"digest"`
+		FullQualifiedDigest string `json:"fully_qualified_digest"`
+		Registry            string `json:"registry"`
+		Repsitory           string `json:"repository"`
+	} `json:"image_summary"`
+	ProvenanceSummary struct {
+		Provenance []struct {
+			Build struct {
+				// TODO: this is untrusted, we should remove it.
+				// IntotoStatement v01IntotoStatement `json:"intotoStatement"`
+			} `json:"build"`
+			Kind        string           `json:"kind"`
+			ResourceUri string           `json:"resourceUri"`
+			Envelope    dsselib.Envelope `json:"envelope"`
+		} `json:"provenance"`
+	} `json:"provenance_summary"`
+}
+
+type GCBProvenance struct {
+	gcloudProv                    *gloudProvenance
+	verifiedIntotoStatementStruct *v01IntotoStatement
+	// Note: may need to support envelope only.
+	// TODO: verified flag?
+}
+
+// go run cli/slsa-verifier/main.go -artifact-path python/secure_package_template-0.2.0-py3-none-any.whl -provenance verifiers/internal/gcb/testdata/test.intoto.jsonl -source github.com/sethmlarson/python-slsa-release-test -builder-id=https://cloudbuild.googleapis.com/GoogleHostedWorker@0.2
+func ProvenanceFromBytes(payload []byte) (*GCBProvenance, error) {
+	var prov gloudProvenance
+	err := json.Unmarshal(payload, &prov)
+	if err != nil {
+		return nil, fmt.Errorf("json.Unmarshal: %w", err)
+	}
+
+	return &GCBProvenance{
+		gcloudProv: &prov,
+	}, nil
+}
+
+func signatureAsRaw(s string) ([]byte, error) {
+	s = strings.ReplaceAll(s, "-", "+")
+	s = strings.ReplaceAll(s, "_", "/")
+	sig, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, err.Error())
+	}
+	return sig, nil
+}
+
+func payloadFromEnvelope(env *dsselib.Envelope) ([]byte, error) {
+	payload, err := base64.StdEncoding.DecodeString(env.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, err.Error())
+	}
+	return payload, nil
+}
+
+func (self *GCBProvenance) isVerified() error {
```

Comment:
recommend changing this to a boolean rather than an error

Reply:
Can do that. Having the `serrors.ErrorNoValidSignature` in one place was easier than having the `if bool return serrors.ErrorNoValidSignature`

Does this change anything?

Reply:
Ah if it's easier to propogate the errors this way, then keep as is -- it was less readable for an `isX` to return an error

Reply:
I can update the name of the funtion. What would our propose?

Reply:
haha, honestly I can't think of one -- `verified()`? would not worry too much because it's internal anyway

Reply:
OK, keeping as-is then. Thanks